### DISCUSSION TITLE
Custom field not appearing on Content-Type Builder.

### DISCUSSION THREAD

**Discussion by SuperMasterBlasterLaser** [strapi/strapi#6351] [Legacy Topics]
- [ ] **I have looked at all the other requests on the Product Board before I submitted this issue**

Hello.

I have created my custom field by this [documentation](https://strapi.io/documentation/3.0.0-beta.x/plugin-development/frontend-field-api.html#plugin-s-front-end-field-api) like this:

    import React, { useEffect, useState } from 'react';

    const Translatable = props => {
  
      console.log('TRANSLATABLE!');
      console.log(props);
  
      return <div>Translatable</div>;
    };

    export default Translatable;

Then I have registered it on `../my-plugin/admin/src/index.js`:

    strapi.registerField({type: 'translatable', Component: Translatable});

Then I run:

    npm run build
    npm run develop

When I open admin and try to add my field to my content type I get this:

![image](https://user-images.githubusercontent.com/8684688/82861447-7cdf0200-9f3e-11ea-9db1-1e78ee9de964.png)

I can't find my custom Field in Content Type Builder. Is it even possible?

There is also one main problem. Where I can describe what kind of data structure this filed should be using on database? Where is backend handling.

In documentation, there are only giant walls of code that describes nothing.

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions)

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests.

**Reply by SuperMasterBlasterLaser** [reply]
OK. Thanks

**Comment by lauriejim** [comment]
Hello! Ping @soupette about this question.
It's his topic :)

**Comment by soupette** [comment]
Hello @SuperMasterBlasterLaser (awesome nickname BTW!).

Regarding the field API as stated in the documentation it is currently only available for the content-manager plugin. We will make some improvements so that the content type builder can display them as well but it will also require some backend modifications which the team don't have the time to do yet (maybe you can make a PR so that it is possible).

Regarding on how to to use your field in the content manager you will need to:

1. Define your entire data structure using the content type builder UI once this is completed you'll need to use the file system
2. Using the file system you'll need to update your `./api/<apiName>/models/<apiName>.settings.json` (it's the same for components) file and add your custom field for instance:

Path: `./api/Address/models/Address.settings.json`

```
{
  "kind": "collectionType",
  "collectionName": "addresses",
  "info": {
    "name": "address",
    "description": ""
  },
  "options": {
    "increments": true,
    "timestamps": ["created_at", "updated_at"],
    "comment": ""
  },
  "attributes": {
    "city": {
      "type": "string",
      "required": true
    },
    "postal_code": {
      "type": "string"
    },
    "coordinates": {
      "type": "geojson" // Custom field that is not in strapi
    }
  }
}
```

Then following the documentation you can create the `Geojson` component so the content manager will use it.

3. By doing so you won't be able to use the content type builder to add new field to your model so make sure to add your custom field at the end.


---

> If you just want to change an input type like the `text` one with a `translatable` one you just need to follow the [documentation](https://strapi.io/documentation/3.0.0-beta.x/guides/registering-a-field-in-admin.html#setup) and do the following ` strapi.registerField({ type: 'text', Component: Translatable });`

I hope this information helps you!

**Reply by mahaveer0496** [reply]
@soupette If we define a custom field like URL which only accepts valid URL, will it work with REST API as well? 
I mean - If in a POST request the field is not a valid URL, will it validate properly?

**Reply by TjaartBroodryk** [reply]
This doesnt seem to work for mongodb as it can't map the type to a valid mongo type. Can we add this capability to the strapi-connector-mongoose package? I'm happy to open a PR for it, if we think its worthwhile? 

Could allow for a `columnType` field - something like the below
```
{
  "Field": {
    "type": "custom type here",
    "columnType": "longtext"
  }
}
```

**Comment by controversial** [comment]
Hey @soupette - if I modify `./api/<apiName>/models/<apiName>.settings.json` as you described, I get the following migration error:

```
error Migration failed
error Error: SQLITE_ERROR: table <apiName> has 5 columns but 6 values were supplied
```

If I modify the `.settings.json` file to have a `type` that's any of the built-in types, the server starts successfully and the migration succeeds, but if I use my custom registered type, the migration fails. Is there somewhere I need to define database config for my custom field type?

**Reply by controversial** [reply]
Update: I found the `columnType` property by reading [BuildDatabaseSchema.js](https://github.com/strapi/strapi/blob/839c51011a07097e5753b5de7e48254e7ef58153/packages/strapi-connector-bookshelf/lib/buildDatabaseSchema.js).

Using like so:
```json
{
  "Field": {
    "type": "custom type here",
    "columnType": "longtext"
  }
}
```
This seems like it solves the issue!

**Reply by Moutaz-homsi** [reply]
Thanks , this solve my issue
